File: src/main/java/org/s16a/mcas/worker/AbstractWorker.java
package org.s16a.mcas.worker;

import com.hp.hpl.jena.rdf.model.Model;
import com.hp.hpl.jena.rdf.model.ModelFactory;
import com.hp.hpl.jena.rdf.model.Property;
import com.hp.hpl.jena.rdf.model.Resource;
import com.hp.hpl.jena.vocabulary.DC;
import com.rabbitmq.client.*;
import org.s16a.mcas.Enqueuer;
import org.s16a.mcas.Hasher;
import org.s16a.mcas.MCAS;
import org.s16a.mcas.util.MediaInfo;

import java.io.*;
import java.net.MalformedURLException;
import java.net.URL;

/**
 * Created by aloeser on 19.05.16.
 */
public abstract class AbstractWorker {

    public final String TASK_QUEUE_NAME;

    public AbstractWorker(Property property) {
        this(property.toString());
    }

    public AbstractWorker(String queueName) {
        TASK_QUEUE_NAME = queueName;
    }


    public static void startWorker(final AbstractWorker worker) throws Exception {
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost("localhost");
        final Connection connection = factory.newConnection();
        final Channel channel = connection.createChannel();


        channel.queueDeclare(worker.TASK_QUEUE_NAME, true, false, false, null);
        System.out.println(" [*] Waiting for messages. To exit press CTRL+C");

        channel.basicQos(1);

        final Consumer consumer = new DefaultConsumer(channel) {
            @Override
            public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body) throws IOException {
                String message = new String(body, "UTF-8");

                System.out.println(" [x] Received '" + message + "'");
                try {
                    worker.processData(message); //usually URL
                } finally {
                    System.out.println(" [x] Done");
                    channel.basicAck(envelope.getDeliveryTag(), false);
                }
            }
        };
        channel.basicConsume(worker.TASK_QUEUE_NAME, false, consumer);
    }

    public Model getNewModel() {
        return ModelFactory.createDefaultModel();
    }

    public Model openModel(String modelFileName) {
        Model model = ModelFactory.createDefaultModel();
        File f = new File(modelFileName);

        if (f.exists()) {
            model.read(modelFileName, "TURTLE");
        }

        return model;
    }

    public void writeModel(Model model, String modelFileName) throws IOException {
        FileWriter out = new FileWriter(modelFileName);
        try {
            model.write(out, "TURTLE");
        } finally {
            try {
                out.close();
            } catch (IOException closeException) {
                // ignore
            }
        }
    }

    public void writeAndMergeModel(Model model, String modelFileName) throws IOException {
        writeModel(model, modelFileName);
        try {
            notifyUpdater(modelFileName);
        } catch (IOException ioe) {
            throw ioe;
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(42);
        }
    }

    public void notifyUpdater(String filename) throws Exception {
        Enqueuer.enqueueForMerge(filename);
    }

    public String getModelFileName(String url) {
        return Hasher.getCacheFilename(url) + "." + TASK_QUEUE_NAME.substring(TASK_QUEUE_NAME.lastIndexOf('/')+1);
    }

    protected abstract void processData(String url) throws IOException;
}

File: python-worker/abstract_worker.py
import os
import pika
import rdflib
from worker_util import get_cache_filename


class AbstractWorker():

    queue_name = None

    def start_worker(self):
        connection = pika.BlockingConnection(
                pika.ConnectionParameters('localhost'))
        channel = connection.channel()

        channel.queue_declare(queue=self.__class__.queue_name, durable=True)
        print(' [*] Waiting for messages. To exit press CTRL+C')
        # channel.basicQos(1)

        def callback(ch, method, properties, body):
            print(' [x] Received %r' % body)
            self.process_data(body)
            ch.basic_ack(delivery_tag=method.delivery_tag)

        channel.basic_consume(callback, queue=self.__class__.queue_name)
        channel.start_consuming()

    def send_to_queue(self, queue_name, body):
        connection = pika.BlockingConnection(
                pika.ConnectionParameters('localhost'))
        channel = connection.channel()

        channel.queue_declare(queue=queue_name, durable=True)

        channel.basic_publish(exchange='',
                              routing_key=queue_name,
                              body=body)

        print(' [x] Sent %r' % body)

        connection.close()

    def get_new_model(self):
        return rdflib.Graph()

    def open_model(self, model_filename):
        model = rdflib.Graph()
        if os.path.isfile(model_filename):
            model_format = rdflib.util.guess_format(model_filename)
            model.parse(model_filename, format=model_format)
        return model

    def notify_update(self, filename):
        self.send_to_queue('update', filename)

    def write_model(self, model, model_filename):
        model.serialize(destination=model_filename, format='turtle')

    def get_model_filename(self, url):
        # url = url.decode('utf-8')
        queue_name_shortened = self.__class__.queue_name.split('/')[-1]
        return get_cache_filename(url) + '.' + queue_name_shortened

    def write_and_merge_model(self, model, filename):
        self.write_model(model, filename)
        self.notify_update(filename)

    def process_data(self, url):
        pass
